internal/controllers/machineset/machineset_controller.go

```diff
@@ -966,6 +1003,156 @@ func (r *Reconciler) getMachineNode(ctx context.Context, cluster *clusterv1.Clus
+	if skipped.Has(clusterv1.MachineSetPreflightCheckAll) {
+		return ctrl.Result{}, nil
+	}
+
+	// If the cluster does not have a control plane reference then there is nothing to do. Return early.
+	if cluster.Spec.ControlPlaneRef == nil {
+		return ctrl.Result{}, nil
+	}
+
+	// Get the control plane object.
+	controlPlane, err := external.Get(ctx, r.Client, cluster.Spec.ControlPlaneRef, cluster.Namespace)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform pre-flight checks: failed to get ControlPlane %s", klog.KRef(cluster.Spec.ControlPlaneRef.Namespace, cluster.Spec.ControlPlaneRef.Name))
+	}
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// If the Control Plane version is not set then we are dealing with a control plane that does not support version
+	// or a control plane where the version is not set. In both cases we cannot perform any preflight checks as
+	// we do not have enough information. Return early.
+	cpVersion, err := contract.ControlPlane().Version().Get(controlPlane)
+	if err != nil {
+		if errors.Is(err, contract.ErrFieldNotFound) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform pre-flight checks: failed to get the version of ControlPlane %s", cpKlogRef)
+	}
+
+	// Run the control plane preflight check.
+	if !skipped.Has(clusterv1.MachineSetPreflightCheckControlPlane) {
+		if result, err := r.controlPlaneStablePreflightCheck(ctx, controlPlane); err != nil || !result.IsZero() {
+			return result, errors.Wrap(err, "failed to perform pre-flight checks")
+		}
+	}
+
+	// Run the kubernetes version skew preflight check.
+	if !skipped.Has(clusterv1.MachineSetPreflightCheckKubernetes) {
+		if result, err := r.kubernetesVersionPreflightCheck(ctx, *cpVersion, ms); err != nil || !result.IsZero() {
+			return result, errors.Wrap(err, "failed to perform pre-flight checks")
+		}
+	}
+
+	// Run the kubeadm version skew preflight check.
+	if !skipped.Has(clusterv1.MachineSetPreflightCheckKubeadm) {
+		if result, err := r.kubeadmVersionPreflightCheck(ctx, *cpVersion, ms); err != nil || !result.IsZero() {
+			return result, errors.Wrap(err, "failed to perform pre-flight checks")
+		}
+	}
+
+	// All pre-flight checks passed.
+	return ctrl.Result{}, nil
+}
+
+func (r *Reconciler) controlPlaneStablePreflightCheck(ctx context.Context, controlPlane *unstructured.Unstructured) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	cpKlogRef := klog.KRef(controlPlane.GetNamespace(), controlPlane.GetName())
+
+	// Check that the control plane is not provisioning.
+	isProvisioning, err := contract.ControlPlane().IsProvisioning(controlPlane)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "failed to perform control plane pre-flight check: failed to check if ControlPlane %s is provisioning", cpKlogRef)
+	}
+	if isProvisioning {
+		log.Info(fmt.Sprintf("Pre-flight check failed: ControlPlane %s is provisioning", cpKlogRef))
```

Comment:
I generally like the log, but I think it would be good if we provide context about what we're trying to do in this moment (e.g. scale up or delete a Machine for remediation).

Maybe we can pass that down here?

Reply:
Are you suggesting something like adding an `action` (or some better name) argument to all the preflight check functions and adjust the logs accordingly?

`action` - could be `Machine Remediation` or `Scale up` and in the logs we will do something like 
```go
log.Info(fmt.Sprintf("Preflight check %q failed while performing %s: ControlPlane %s is provisioning", clusterv1.MachineSetPreflightCheckControlPlaneStable, action, cpKlogRef))
```
(will also include an if else block to handle the log incase `action == ""`)

Another option would be to pass a a modifed logger in the context that has a key-value pair which would specify the context of the preflight checks.

Reply:
The first. Although it would be nice to flip the order and mention what we're doing first

Reply:
Changed it to add the `action` parameter. PTAL.